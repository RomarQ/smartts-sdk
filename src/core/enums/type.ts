export enum TypeAtom {
    record = 'record',
    variant = 'variant',
    chain_id = 'chain_id',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    option = 'option',
    map = 'map',
    big_map = 'bigmap',
    signature = 'signature',
    operation = 'operation',
    never = 'never',
    tuple = 'tuple',
    lambda = 'lambda',
    ticket = 'ticket',
    sapling_state = 'sapling_state',
    contract = 'contract',
    sapling_transaction = 'sapling_transaction',
}

export default TypeAtom;
