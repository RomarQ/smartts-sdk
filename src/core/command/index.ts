export * from './controls';
export * from './variables';
export * from './function';
