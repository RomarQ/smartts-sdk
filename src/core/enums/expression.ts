export enum ExpressionAtom {
    // Math
    add = 'add',
    mul = 'mul',
    sub = 'sub',
    truediv = 'truediv',
    // Comparison
    eq = 'eq',
    neq = 'neq',
    lt = 'lt',
    gt = 'gt',
    le = 'le',
    ge = 'ge',
    //
    items = 'items',
    update_map = 'update_map',
    getItem = 'getItem',
    getItemDefault = 'getItemDefault',
    getItemMessage = 'getItemMessage',
    attr = 'attr',
    params = 'params',
    getLocal = 'getLocal',
    iter = 'iter',
    // Blockchain property
    amount = 'amount',
    balance = 'balance',
    chain_id = 'chain_id',
    level = 'level',
    now = 'now',
    self = 'self',
    self_address = 'self_address',
    sender = 'sender',
    source = 'source',
    total_voting_power = 'total_voting_power',
    voting_power = 'voting_power',
    // Operation
    operations = 'operations',
    transfer = 'transfer',
    set_delegate = 'set_delegate',
    create_contract = 'create_contract',
    // List instructions
    cons = 'cons',
    //
    contract = 'contract',
    // Variant
    openVariant = 'openVariant',
    isVariant = 'isVariant',
    cases_arg = 'cases_arg',
    variant_arg = 'variant_arg',
}

export default ExpressionAtom;
