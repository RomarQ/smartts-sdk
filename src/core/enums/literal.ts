export enum LiteralAtom {
    record = 'record',
    variant = 'variant',
    chain_id_cst = 'chain_id_cst',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    Some = 'Some',
    None = 'None',
    Left = 'Left',
    Right = 'Right',
    map = 'map',
    big_map = 'big_map',
    signature = 'signature',
    tuple = 'tuple',
    ticket = 'ticket',
    sapling_state = 'sapling_empty_state',
}

export default LiteralAtom;
