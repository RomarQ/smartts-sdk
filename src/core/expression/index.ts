export * from './literal';
export * from './expression';
export * from './logic';
