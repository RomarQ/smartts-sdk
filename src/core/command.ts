import Utils, { LineInfo } from '../misc/utils';
import { IExpression } from '../typings/expression';
import { IStatement } from '../typings/statement';

class C_DefineLocal implements IStatement {
    constructor(public name: string, public value: unknown, public mutable: boolean, public line = new LineInfo()) {}

    toString() {
        return `(defineLocal "${this.name}" ${this.value} ${Utils.capitalizeBoolean(this.mutable)} ${this.line})`;
    }
}
export const DefineVar = (name: string, value: unknown, mutable = true, line = new LineInfo()) =>
    new C_DefineLocal(name, value, mutable, line);

class C_Verify implements IStatement {
    constructor(public condition: IExpression, public errorMsg: IExpression, public line = new LineInfo()) {}

    toString() {
        return `(verify ${this.condition} ${this.errorMsg} ${this.line})`;
    }
}
export const Require = (condition: IExpression, errorMsg: IExpression, line = new LineInfo()) =>
    new C_Verify(condition, errorMsg, line);

class C_SetValue implements IStatement {
    name = 'set';
    constructor(public source: unknown, public value: unknown, public line = new LineInfo()) {}

    toString() {
        return `(${this.name} ${this.source} ${this.value} ${this.line})`;
    }
}
export const SetValue = (source: unknown, value: unknown, line = new LineInfo()) => new C_SetValue(source, value, line);

class IfStatment implements IStatement {
    #condition: IExpression;
    #thenStatements: IStatement[];
    #elseStatements: IStatement[];
    #line: LineInfo;

    constructor(
        condition: IExpression,
        line = new LineInfo(),
        thenStatements?: IStatement[],
        elseStatements?: IStatement[],
    ) {
        this.#condition = condition;
        this.#thenStatements = thenStatements || [];
        this.#elseStatements = elseStatements || [];
        this.#line = line;
    }

    public Then(statements: IStatement[]) {
        this.#thenStatements = statements;
        return this;
    }

    public Else(statements: IStatement[]) {
        this.#elseStatements = statements;
        return this;
    }

    private ifBlock() {
        return `(ifBlock ${this.#condition} (${this.#thenStatements.join(' ')}) ${this.#line})`;
    }

    private elseBlock() {
        return `(elseBlock (${this.#elseStatements.join(' ')}))`;
    }

    [Symbol.toPrimitive]() {
        let expression = this.ifBlock();
        if (this.#elseStatements.length) {
            expression = `${expression} ${this.elseBlock()}`;
        }
        return expression;
    }
}
export const If = (condition: IExpression, thenStatements?: IStatement[], elseStatements?: IStatement[]) =>
    new IfStatment(condition, new LineInfo(), thenStatements, elseStatements);

const Commands = {
    DefineVar,
    Require,
    SetValue,
    If,
};

export default Commands;
