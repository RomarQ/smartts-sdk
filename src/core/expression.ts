import type { IExpression } from '../typings/expression';
import type { ILiteral } from '../typings/literal';
import type { IStatement } from '../typings/statement';
import type { IType } from '../typings/type';
import { GetProperty, LambdaArgument } from '../expression/variables';
import { capitalizeBoolean, LineInfo, parenthesis } from '../misc/utils';
import ValueAtom from './enums/literal';
import TypeAtom from './enums/type';

export class Expression<T extends ValueAtom> implements IExpression<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;
    private args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return GetProperty(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}

export class LiteralExpression<T extends ValueAtom> implements ILiteral<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;

    constructor(private name: T, private values: (number | string | boolean | IExpression)[], private line: LineInfo) {}

    toString() {
        switch (this.name) {
            case ValueAtom.unit:
                return parenthesis(this.name);
            case ValueAtom.nat:
            case ValueAtom.int:
            case ValueAtom.mutez:
            case ValueAtom.bytes:
            case ValueAtom.bool:
            case ValueAtom.string:
            case ValueAtom.chain_id_cst:
            case ValueAtom.timestamp:
            case ValueAtom.bls12_381_fr:
            case ValueAtom.bls12_381_g1:
            case ValueAtom.bls12_381_g2:
            case ValueAtom.key:
            case ValueAtom.key_hash:
            case ValueAtom.signature:
            case ValueAtom.address:
                return `(literal (${this.name} ${this.values.join(' ')}) ${this.line})`;
            case ValueAtom.list:
            case ValueAtom.set:
            case ValueAtom.tuple:
                return `(${this.name} ${this.line} ${this.values.join(' ')})`;
            case ValueAtom.Some:
            case ValueAtom.None:
                return `(${ValueAtom.variant} "${this.name}" ${this.values.join(' ') || `(${ValueAtom.unit})`} ${
                    this.line
                })`;
            default:
                return `(${this.name} ${this.values.join(' ')} ${this.line})`;
        }
    }
}

export class RecordLiteral implements ILiteral<ValueAtom.record> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as ValueAtom.record;

    constructor(private fields: Record<string, IExpression>, private line: LineInfo) {}

    private buildFields = (fields: Record<string, IExpression>) => {
        return Object.entries(fields).map(([field, expr]) => `(${field} ${expr.toString()})`);
    };

    toString(): string {
        return `(${TypeAtom.record} ${this.line} ${this.buildFields(this.fields).join(' ')})`;
    }
}

export class MapLiteral<T extends ValueAtom.map | ValueAtom.big_map> implements ILiteral<T> {
    _isExpression = true as const;
    _type = {} as T;
    type = {} as IType;

    constructor(
        private prim: T,
        private rows: IExpression[][],
        keyType: IType,
        valueType: IType,
        private line: LineInfo,
    ) {}

    private buildEntry = ([key, value]: IExpression[]) => {
        return `(${key.toString()} ${value.toString()})`;
    };

    toString() {
        return `(${this.prim} ${this.line} ${this.rows.map(this.buildEntry).join(' ')})`;
    }
}

export class LambdaLiteral implements ILiteral<ValueAtom.lambda> {
    _isExpression = true as const;
    // Used for type checking
    _type = ValueAtom.lambda as const;
    type = {} as unknown as IType;
    static idCounter = 0;
    private identifier: number;
    private withStorage?: 'read-write' | 'read-only';
    private withOperations = false;

    constructor(
        private statements: IStatement[],
        private inType: IType,
        private argumentName: string,
        private line: LineInfo,
    ) {
        this.identifier = ++LambdaLiteral.idCounter;
    }

    public get id(): Readonly<number> {
        return this.identifier;
    }

    public setInputType(type: IType) {
        this.inType = type;
        return this;
    }

    public code(buildStatements: (arg: IExpression) => IStatement[]) {
        const param = LambdaArgument(this.argumentName, this.inType, this.id, this.line);
        this.statements = buildStatements(param);
        return this;
    }

    toString() {
        return `(${ValueAtom.lambda} ${this.id} ${this.withStorage || 'None'} ${capitalizeBoolean(
            this.withOperations,
        )} "${this.argumentName}" ${this.line} (${this.statements.join(' ')}))`;
    }
}
