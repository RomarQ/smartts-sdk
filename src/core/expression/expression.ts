import { proxy } from '../../misc/proxy';
import { LineInfo } from '../../misc/utils';
import { IExpression } from '../../typings/expression';
import { ILiteral } from '../../typings/literal';
import { IType } from '../../typings/type';

export class Expression implements IExpression {
    _isExpression = true as const;
    args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string | ILiteral<unknown>)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return Expression.getAttr(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    static getAttr = (from: IExpression, attr: string, line = new LineInfo()) =>
        proxy(new Expression('attr', from, `"${attr}"`, line), Expression.proxyHandler);

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}
