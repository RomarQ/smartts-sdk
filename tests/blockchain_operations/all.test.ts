import { None, Some } from '../../src/core/literal';
import {
    GetChain_id,
    Contract,
    EntryPoint,
    GetTimestamp,
    GetLevel,
    GetSender,
    GetSource,
    TOption,
    TUnit,
    TAddress,
    TTimestamp,
    TChain_id,
    TNat,
} from '../../src/core';
import { ContractStorage } from '../../src/core/expression';
import { SetValue } from '../../src/core/command';
import { verifyMichelsonOutput } from '../util';
import { IType } from '../../src/typings/type';
import { IExpression } from '../../src/typings/expression';

const verify = (testName: string, type: IType, operation: IExpression) => {
    it(testName, () => {
        const contract = new Contract()
            .setStorageType(TOption(type))
            .setStorage(None())
            .addEntrypoint(
                new EntryPoint('ep1').inputType(TUnit()).code(() => [SetValue(ContractStorage(), Some(operation))]),
            )
            .toString();

        expect(contract).toMatchSnapshot();
        verifyMichelsonOutput(contract);
    });
};

describe('Test blockchain operations', () => {
    verify('Get Sender', TAddress(), GetSender());
    verify('Get Source', TAddress(), GetSource());
    verify('Get Block Level', TNat(), GetLevel());
    verify('Get Timestamp', TTimestamp(), GetTimestamp());
    verify('Get Chain ID', TChain_id(), GetChain_id());
});
