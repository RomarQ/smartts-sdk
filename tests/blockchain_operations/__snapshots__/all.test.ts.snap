// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test blockchain operations AMOUNT 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option \\"mutez\\"))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (amount) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations AMOUNT 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"mutez\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"AMOUNT\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test blockchain operations BALANCE 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option \\"mutez\\"))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (balance) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations BALANCE 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"mutez\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"BALANCE\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test blockchain operations CHAIN_ID 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option \\"chain_id\\"))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (chain_id) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations CHAIN_ID 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"chain_id\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CAR\\"
                },
                {
                    \\"prim\\": \\"CHAIN_ID\\"
                },
                {
                    \\"prim\\": \\"SWAP\\"
                },
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test blockchain operations LEVEL 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option \\"nat\\"))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (level) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations LEVEL 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"nat\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CAR\\"
                },
                {
                    \\"prim\\": \\"LEVEL\\"
                },
                {
                    \\"prim\\": \\"SWAP\\"
                },
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test blockchain operations NOW 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option \\"timestamp\\"))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (now) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations NOW 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"timestamp\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"NOW\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test blockchain operations SELF 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option (contract \\"unit\\")))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (self \\"\\" (\\"all.test.ts\\" 40)) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations SELF 2`] = `"\\"import: params type yet unknown\\""`;

exports[`Test blockchain operations SELF_ADDRESS 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option \\"address\\"))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (self_address) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations SELF_ADDRESS 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"address\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"SELF_ADDRESS\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test blockchain operations SENDER 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option \\"address\\"))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (sender) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations SENDER 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"address\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"SENDER\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test blockchain operations SOURCE 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option \\"address\\"))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (source) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations SOURCE 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"address\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"SOURCE\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test blockchain operations TOTAL_VOTING_POWER 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option \\"nat\\"))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (total_voting_power) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations TOTAL_VOTING_POWER 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"nat\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CAR\\"
                },
                {
                    \\"prim\\": \\"TOTAL_VOTING_POWER\\"
                },
                {
                    \\"prim\\": \\"SWAP\\"
                },
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test blockchain operations VOTING_POWER 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 27))
            storage (variant \\"None\\" (unit) (\\"all.test.ts\\" 29))
            storage_type ((option \\"nat\\"))
            messages ((ep1 True False False True (\\"all.test.ts\\" 31) ((set_type (params (\\"all.test.ts\\" 31)) \\"unit\\" (\\"all.test.ts\\" 31)) (set (data) (variant \\"Some\\" (voting_power (literal (key_hash tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN) (\\"all.test.ts\\" 40)) (\\"all.test.ts\\" 40)) (\\"all.test.ts\\" 31)) (\\"all.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 27))
        )
        "
`;

exports[`Test blockchain operations VOTING_POWER 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"nat\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"DROP\\"
                },
                {
                    \\"prim\\": \\"PUSH\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"key_hash\\"
                        },
                        {
                            \\"string\\": \\"tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"VOTING_POWER\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;
